contracts/token/ERC721/ERC721.sol

```diff
@@ -331,8 +331,24 @@ contract ERC721 is Context, ERC165, IERC721 {
             return true;
         }
 
-        bytes4 retval = IERC721Receiver(to).onERC721Received(_msgSender(), from, tokenId, _data);
-        return (retval == _ERC721_RECEIVED);
+        bytes memory payload = abi.encodeWithSignature(
+            "onERC721Received(address,address,uint256,bytes)",
+            msg.sender,
+            from,
+            tokenId,
+            _data
+            );
+
+            //solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory returndata) = to.call(payload);
+        if (!success) {
+            if (returndata.length > 0){
+                revert(string(returndata));
```

Comment:
From reading the Solidity documentation, it seems that both `revert` and `require` abi-encode their arguments as if they were a call to an `Error(string)` function.

If I understand correctly, this means that we're "wrapping" the revert reason twice, since `returndata` is already encoded here. I think we have to drop to assembly and use the `revert` opcode directly to avoid this re-encoding and to bubble up the value as-is.

Let me know if this was clear @NoopurShinde.